app/src/main/java/org/stacks/app/ui/auth/signup/ChooseUsernameActivity.kt

```diff
@@ -42,22 +56,71 @@ class ChooseUsernameActivity : BaseActivity() {
 
         viewModel
             .openNewAccountScreen()
+            .onEach { startActivity(WelcomeActivity.getIntent(this)) }
+            .launchIn(lifecycleScope)
+
+        viewModel
+            .loading()
+            .onEach { loading ->
+                loadingSpinner.isVisible = loading
+                continueBtn.isVisible = !loading
+                chooseUsernameLayout.isVisible = !loading
+            }
+            .launchIn(lifecycleScope)
+
+        viewModel
+            .sendAuthResponse()
             .onEach {
-                startActivity(WelcomeActivity.getIntent(this))
+
+                if (signUp) {
+                    startActivity(
+                        WelcomeActivity.getIntent(
+                            this,
+                            it.appName,
+                            it.redirectURL,
+                            it.authResponseToken
+                        )
+                    )
+                } else {
+                    sendAuthResponse(it)
+                }
             }
             .launchIn(lifecycleScope)
 
         viewModel
             .errors()
-            .onEach {
-                outlinedTextField.error = getString(R.string.error)
+            .onEach { error ->
+                outlinedTextField.error = when (error) {
+                    ChooseUsernameViewModel.Error.UnavailableUsername -> getString(R.string.invalidUsername)
+                    ChooseUsernameViewModel.Error.SignUpError -> getString(R.string.error)
+                }
+
             }
             .launchIn(lifecycleScope)
     }
 
+    private fun sendAuthResponse(authResponse: AuthResponse) {
+        val uri = Uri.parse(authResponse.redirectURL)
+            .buildUpon()
+            .appendQueryParameter(IdentitiesActivity.AUTH_RESPONSE, authResponse.authResponseToken)
+            .build()
+
+        startActivity(
+            Intent(
+                Intent.ACTION_VIEW,
+                uri
+            )
+        )
+
+        finishAffinity()
```

Comment:
You are doing this is a lot of places, maybe some helper could be used, at least to construct the intent.